model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java

```diff
@@ -270,25 +272,16 @@ private UserSessionEntity getUserSessionEntity(RealmModel realm, String id, bool
     }
 
 
-    protected List<UserSessionModel> getUserSessions(RealmModel realm, Predicate<Map.Entry<String, SessionEntityWrapper<UserSessionEntity>>> predicate, boolean offline) {
+    protected Stream<UserSessionModel> getUserSessionsStream(RealmModel realm, Predicate<Map.Entry<String, SessionEntityWrapper<UserSessionEntity>>> predicate, boolean offline) {
         Cache<String, SessionEntityWrapper<UserSessionEntity>> cache = getCache(offline);
-
         cache = CacheDecorators.skipCacheLoaders(cache);
 
-        Stream<Map.Entry<String, SessionEntityWrapper<UserSessionEntity>>> cacheStream = cache.entrySet().stream();
-
-        List<UserSessionModel> resultSessions = new LinkedList<>();
-
-        Iterator<UserSessionEntity> itr = cacheStream.filter(predicate)
+        // return a stream that 'wraps' the infinispan cache stream so that the cache stream's elements are read one by one
+        // and then filtered/mapped locally to avoid serialization issues when trying to manipulate the cache stream directly.
+        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(cache.entrySet().stream().iterator(), 0), true)
```

Comment:
Would the following work as well?
```suggestion
        return StreamSupport.stream(cache.entrySet().stream().spliterator(), true)
```

Reply:
Yes, it does. Not sure why I haven't tried it first.